[MXNET -1004] Poisson NegativeLog Likelihood loss

gaurav-gireesh opened the pull request:
## Description ##
The function computes the poisson's negative log likelihood loss that serves as measurement of the loss accrued from a Poisson regression prediction made by the model.


## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)

- [x] Code is well-documented: 

- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] PoissonNLLLoss computation function tests, (and when applicable, API doc)


## Comments ##
- If this change is a backward incompatible change, why must this change be made.

 - Below is the link for pytorch's current implementation:
    [PyTorch Implementation] . 
    (https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py)

vandanavk commented:
@mxnet-label-bot [pr-work-in-progress]

anirudhacharya commented:
@gaurav-gireesh what is in progress for this PR? The PR description says that you have completed all the code changes and all the checkboxes are ticked.

roywei commented on tests/python/unittest/test_loss.py:
```diff
@@ -185,27 +185,28 @@ def test_l1_loss():
 def test_ctc_loss():
     loss = gluon.loss.CTCLoss()
-    l = loss(mx.nd.ones((2,20,4)), mx.nd.array([[1,0,-1,-1],[2,1,1,-1]]))
+    l = loss(mx.nd.ones((2, 20, 4)), mx.nd.array([[1, 0, -1, -1], [2, 1, 1, -1]]))
```
As discussed, please address pylint issue in another PR, it should be going to master first before this PR

roywei commented on tests/python/unittest/test_loss.py:
```diff
@@ -342,12 +343,45 @@ def test_triplet_loss():
     assert mod.score(data_iter, eval_metric=mx.metric.Loss())[0][1] < 0.05
 
+@with_seed()
+def test_poisson_nllloss():
```
Please separate into 3 unit tests clearly stating what condition are you testing.

gaurav-gireesh replied:
@Lai Yes, I have separated the use cases commenting on what exactly is being tested. I have not split into different functions so as to keep consistency with other loss functions tests.
Thank you for your comment.

roywei commented on python/mxnet/gluon/loss.py:
```diff
@@ -706,3 +707,64 @@ def hybrid_forward(self, F, pred, positive, negative):
+        if self._compute_full:
+            stirling_factor = target * F.log(target)- target + 0.5 * F.log(2 * target * np.pi)
+            target_gt_1 = target > 1
+            stirling_factor *= target_gt_1
```
Add a comment explaining why we need this

gaurav-gireesh replied:
Thanks!
Yes, I have mentioned the use of the parameter 'compute_full' in function docstring, which explains why the calculation is needed.

roywei commented:
> @gaurav-gireesh what is "in progress" for this PR? The PR description says that you have completed all the code changes and all the checkboxes are ticked.

@anirudhacharya  there was an issue with CI and pylin, we are looking into it.

gaurav-gireesh commented:
> @gaurav-gireesh what is "in progress" for this PR? The PR description says that you have completed all the code changes and all the checkboxes are ticked.

@anirudhacharya It was a miss Anirudha. Corrected that and added a few test cases for Stirling approximation too.

anirudhacharya commented on python/mxnet/gluon/loss.py:
```diff
@@ -706,3 +707,65 @@ def hybrid_forward(self, F, pred, positive, negative):
+    from_logits : boolean, default True
+        indicating whether log(predicted) value has already been computed. If True, the loss is computed as
+        :math:`\exp(\text{pred}) - \text{target} * \text{pred}`, and if False, then loss is computed as
+        :math:`\text{pred} - \text{target} * \log(\text{pred}+\text{epsislon})`.The default value
```
these definitions seem to be at odds with the definition given above in line 718

gaurav-gireesh replied:
Yes the formulae are different. Following are some points:
1)  The factorial term gets dropped in most calculations. However, there is a way to approximate it using Stirling factor/approximation. We can use **_compute_full_** flag to be set to True to incorporate this approximation in our loss value computation,
2) from_logits flag set to true or false can render the formula to change depending upon the prediction is already logged ( log pred) or not.

anirudhacharya commented on tests/python/unittest/test_loss.py:
```diff
@@ -348,6 +348,44 @@ def test_triplet_loss():
+    # 1) Testing for flag logits = True
+    brute_loss = np.mean(np.exp(pred.asnumpy()) - target.asnumpy() * pred.asnumpy())
+    loss_withlogits = Loss(pred, target)
+    assert_almost_equal(brute_loss, loss_withlogits.asscalar())
```
why not test this loss function using the Module API with `module.fit` and `module.score`

gaurav-gireesh replied:
Thank you for the suggestion. It is a valid point however, getting a synthetic data set to behave in a way that Random input X's are correlated to a target ~ PoissonDistribution(target belonging to a Poisson Distribution) is difficult to gather. However, I have tried with random data sets and trained the model but the loss gradient was not observed. This could be incrementally contributed if I come across such a dataset which shows the loss gradient with epochs as other models like LogisticRegression, Linear Regression and so on.
The formula to compute a loss value however, is implemented in the function which can be tested by raw calculations. This is also something that we see in unit tests of loss functions such as : test_bce_loss(Binary Cross Entropy).

gaurav-gireesh replied:
Unit test added.

gaurav-gireesh commented on python/mxnet/gluon/loss.py:
```diff
@@ -706,3 +707,65 @@ def hybrid_forward(self, F, pred, positive, negative):
+    PoissonNLLLoss measures the loss accrued from a poisson regression prediction made by the model.
+
+    .. math::
+        L = \text{pred} - \text{target} * \log(\text{pred}) +\log(\text{target!})
```
@anirudhacharya The "!" is for factorial of the target value. The formula to calculate the probability of the target value with the mean of the poisson distribution contains a factorial term in the denominator which is approximated by computing the Stirling approximation. Taking a log of the formula reduces to the form mentioned in the documentation.
Some wikipedia pages that I have found helpful are:
[Poisson regression](https://en.wikipedia.org/wiki/Poisson_regression) and
[Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution).
Also, I have attached a link for reference to PyTorch's implementation of the loss function in the description of the PR.

sandeep-krishnamurthy commented on python/mxnet/gluon/loss.py:
```diff
@@ -23,8 +23,8 @@
            'KLDivLoss', 'CTCLoss', 'HuberLoss', 'HingeLoss',
-           'SquaredHingeLoss', 'LogisticLoss', 'TripletLoss']
-
+           'SquaredHingeLoss', 'LogisticLoss', 'TripletLoss', 'PoissonNLLLoss']
```
nit: please add line space

gaurav-gireesh replied:
Added it. Thanks for pointing out.

sandeep-krishnamurthy commented on tests/python/unittest/test_loss.py:
```diff
@@ -348,6 +348,44 @@ def test_triplet_loss():
+        assert_almost_equal(np_loss_no_logits, loss_no_logits.asscalar())
+
+    #3) Testing for Sterling approximation
+    np_pred = np.random.uniform(1, 5, (2, 3))
```
Please add tests for hybridized version as well.

gaurav-gireesh replied:
> Please add tests for hybridized version as well.

Added the unit test for it.

gaurav-gireesh commented:
@mxnet-label-bot [pr-awaiting-review]

gaurav-gireesh commented:
@sandeep-krishnamurthy This PR is awaiting review, so I request you to change the label.
Thanks!
